sdk/compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs

```diff
@@ -536,12 +540,70 @@ generatePackageMap version mbProjRoot userPkgDbs = do
+        , T.pack $ LF.renderVersion $ optDamlLfVersion opts
+        , ") does not support Smart Contract Upgrades"
+        ]
+  , justIf (not $ LF.packageLfVersion mainPkg `LF.supports` LF.featurePackageUpgrades) $
+      ideErrorPretty file $ mconcat
+        [ "Upgraded package LF Version ("
+        , T.pack $ LF.renderVersion $ LF.packageLfVersion mainPkg
+        , ") does not support Smart Contract Upgrades"
+        ]
+  , justIf (optDamlLfVersion opts `lfVersionLt` LF.packageLfVersion mainPkg) $
+      ideErrorPretty file $ mconcat
+        [ "Upgraded package LF Version ("
+        , T.pack $ LF.renderVersion $ LF.packageLfVersion mainPkg
+        , ") cannot be higher than main package LF Version ("
+        , T.pack $ LF.renderVersion $ optDamlLfVersion opts
+        , ")"
+        ]
+  ] ++ case LF.packageMetadata mainPkg of
+    Nothing -> [Just $ ideErrorPretty file ("Upgraded DAR does not contain metadata" :: T.Text)]
+    Just meta ->
+      [ justIf (optMbPackageName opts /= Just (LF.packageName meta)) $
+          ideErrorPretty file $ T.unlines
+            [ "Upgraded package must have the same package name as main package."
+            , "Expected " <> maybe "<unknown>" LF.unPackageName (optMbPackageName opts)
+            , "Got " <> LF.unPackageName (LF.packageName meta)
+            ]
+      , justIf (optMbPackageVersion opts == Just (LF.packageVersion meta)) $
+          ideErrorPretty file $ mconcat
+            [ "Upgraded package cannot have the same package version as main package ("
+            , LF.unPackageVersion $ LF.packageVersion meta
+            , ")"
+            ]
+      , justIf (maybe False (`packageVersionLt` LF.packageVersion meta) $ optMbPackageVersion opts) $
+          ideErrorPretty file $ mconcat
+            [ "Upgraded package cannot have a higher package version ("
+            , LF.unPackageVersion $ LF.packageVersion meta
+            , ") than main package ("
+            , LF.unPackageVersion $ fromJust $ optMbPackageVersion opts
+            , ")"
+            ]
+      ]
+  where
+    justIf :: Bool -> a -> Maybe a
+    justIf cond val = guard cond >> Just val
+
+    packageVersionLt :: LF.PackageVersion -> LF.PackageVersion -> Bool
+    packageVersionLt = (<) `on` fmap (read @Int) . splitWhen (=='.') . T.unpack . LF.unPackageVersion
+
+    -- We assume major will always be 1 (or at least the same), since Version.supportedOutputVersions does not contain any 2. versions
+    -- TODO: Is this acceptable?
+    lfVersionLt :: LF.Version -> LF.Version -> Bool
+    lfVersionLt = (<) `on` LF.versionMinor
+
 extractUpgradedPackageRule :: Options -> Rules ()
 extractUpgradedPackageRule opts = do
   defineNoFile $ \ExtractUpgradedPackage ->
     case uiUpgradedPackagePath (optUpgradeInfo opts) of
       Nothing -> pure Nothing
-      Just path -> use ExtractUpgradedPackageFile (toNormalizedFilePath' path)
+      Just path -> Just <$> use_ ExtractUpgradedPackageFile (toNormalizedFilePath' path)
```

Comment:
We *could* rewrite this case as `use_ ExtractUpgradedPackageFile . toNormalizedFilePath =<< uiUpgradedPackagePath (optUpgradeInfo opts)` I think, but that's a bit of a mouthful, up to you if you want to stick to case statements over combinators.

Reply:
Can you? `uiUpgradedPackagePath (optUpgradeInfo opts)` is a `Maybe FilePath`, and you'd be binding it with `FilePath -> Action ...` 
I think you'd need traverse

Reply:
I used traverse :)